L1Trigger/L1THGCal/src/be_algorithms/HGCalShowerShape.cc

```diff
@@ -0,0 +1,554 @@
+    for(int i=0;i<ntc;i++){
+        if(phi.at(i)>0) phi_sum+=energy.at(i)*phi.at(i);
+        else phi_sum+=energy.at(i)*(phi.at(i)+2*TMath::Pi());
+        w_sum+=energy.at(i);
+    }
+    float phi_mean=phi_sum/w_sum;
+    if(phi_mean>TMath::Pi()) phi_mean-=2*TMath::Pi();
+    //compute weighted eta RMS
+    float deltaphi2_sum=0;
+    for(int i=0;i<ntc;i++){
+        float deltaPhi=std::abs(phi.at(i)-phi_mean);
+        if (deltaPhi>TMath::Pi()) deltaPhi=2*TMath::Pi()-deltaPhi;
+        deltaphi2_sum+=energy.at(i)*pow(deltaPhi,2);
+    }        
+    float phi_RMS=0;
+    if (w_sum!=0) phi_RMS=deltaphi2_sum/w_sum;
+    float Spp=sqrt(phi_RMS);
+    return Spp;
+    
+}
+
+
+
+const float HGCalShowerShape::sigmaZZ(const std::vector<float> energy, const std::vector<float> z){
+
+    int ntc=energy.size();
+    //compute weighted eta mean
+    float z_sum=0;
+    float w_sum=0; //here weight is energy
+    for(int i=0;i<ntc;i++){
+        z_sum+=energy.at(i)*z.at(i);
+        w_sum+=energy.at(i);
+    }
+    float z_mean=z_sum/w_sum;
+    //compute weighted eta RMS
+    float deltaz2_sum=0;
+    for(int i=0;i<ntc;i++) deltaz2_sum+=energy.at(i)*pow((z.at(i)-z_mean),2);
+    float z_RMS=0;
+    if (w_sum!=0) z_RMS=deltaz2_sum/w_sum;
+    float Szz=sqrt(z_RMS);
+    return Szz;
+    
+}
+
+
+const int HGCalShowerShape::showerLength(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    int firstLayer=999;
+    int lastLayer=-999;
+    int layer=999;
+
+    int nLayersEE=28;
+   
+    for(const auto& clu : clustersPtrs){
+
+        if(clu->subdetId()==HGCEE) layer=clu->layer();//EE
+        if(clu->subdetId()==HGCHEF) layer=clu->layer()+nLayersEE;//FH
+
+        if(layer<firstLayer) firstLayer=layer;
+        if(layer>lastLayer) lastLayer=layer;
+
+    }
+
+    int nLayers=lastLayer-firstLayer+1;
+
+    return nLayers;
+
+}
+
+
+const int HGCalShowerShape::firstLayer(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    int firstLayer=999;
+    int layer=999;
+
+    int nLayersEE=28;
+
+    for(const auto& clu : clustersPtrs){
+
+        if(clu->subdetId()==HGCEE) layer=clu->layer();//EE
+        if(clu->subdetId()==HGCHEF) layer=clu->layer()+nLayersEE;//FH
+
+        if(layer<firstLayer) firstLayer=layer;
+
+    }
+
+    return firstLayer;
+
+}
+
+
+
+
+const float HGCalShowerShape::sigmaEtaEtaTot(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    std::vector<float> tc_energy ; 
+    std::vector<float> tc_eta ;
+
+    for(const auto& clu : clustersPtrs){
+
+        const edm::PtrVector<l1t::HGCalTriggerCell> triggerCells = clu->constituents();
+        unsigned int ncells = triggerCells.size();
+
+        for(unsigned int itc=0; itc<ncells;itc++){
+
+            l1t::HGCalTriggerCell thistc = *triggerCells[itc];
```

Comment:
Better to use directly the pointer instead of copying the trigger cell. Or even better, use a range-based loop:
```cpp
for(const auto& tc : triggerCells)
{
  tc_energy.emplace_back(tc->energy());
  tc_eta.emplace_back(tc->eta());
}
```